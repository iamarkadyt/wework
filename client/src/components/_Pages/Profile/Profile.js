import React from 'react'
import { connect } from 'react-redux'
import { Route } from 'react-router-dom'
import './Profile.css'

import ProfileView from '../../ProfileView/ProfileView'
import UpdateProfile from '../../_Forms/UpdateProfile/UpdateProfile'
import AddEdu from '../../_Forms/AddEdu/AddEdu'
import AddExp from '../../_Forms/AddExp/AddExp'
import Modal from '../../Modal/Modal'
import { fetchUsersProfile, deleteProfile } from '../../../state/actions/profileActions'

class Profile extends React.Component {
    render() {
        const {
            history,
            match,
            fetchUsersProfile,
            deleteProfile,
            profile: profileData,
            errors: {
                noprofile
            }
        } = this.props

        const baseUrl = match.url || ''

        let content = <div>
            <h2>Please wait, loading...</h2>
        </div>
        if (noprofile) {
            content = (
                <div className="Profile-content">
                    <Route path={baseUrl} exact render={() => (
                        <Modal
                            question="You don't yet have a profile. Would you like to create one now?"
                            onConfirm={() => history.push(`${baseUrl}/create-profile`)}
                            onDismiss={() => history.push('/feed')}
                            actionColor="green" />
                    )} />
                    <Route path={`${baseUrl}/create-profile`} render={props => (
                        <UpdateProfile {...props} />
                    )} />
                </div>
            )
        } else if (profileData) {
            content = (
                <div className="Profile-content">
                    <Route path={baseUrl} render={props => (
                        <ProfileView {...props} profile={profileData} />
                    )} />
                    <Route path={`${baseUrl}/update-profile`} render={props => (
                        <UpdateProfile {...props} profile={profileData} />
                    )} />
                    <Route path={`${baseUrl}/add-experience`} render={props => (
                        <AddExp {...props} />
                    )} />
                    <Route path={`${baseUrl}/add-education`} render={props => (
                        <AddEdu {...props} />
                    )} />
                    <Route path={`${baseUrl}/delete-profile`} render={() => (
                        <Modal
                            question="Are you sure you want to delete your profile?"
                            onConfirm={() => deleteProfile(() => history.push(baseUrl))}
                            onDismiss={() => history.goBack()}
                            actionColor="red" />
                    )} />
                </div>
            )
        } else {
            fetchUsersProfile()
        }

        return (
            <div className="Profile-container">
                {content}
            </div>
        )
    }
}

export default connect(state => ({
    profile: state.profile,
    errors: state.err
}), { fetchUsersProfile, deleteProfile })(Profile)