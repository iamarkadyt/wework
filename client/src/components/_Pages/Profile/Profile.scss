.Profile-container {
    display: flex;
    justify-content: center;
    padding: 1rem;
    .Profile-content {
        width: 50rem;
    }
}