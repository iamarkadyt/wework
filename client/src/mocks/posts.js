export const mockPost = {
    "_id": "c717a50ac29bacb2ee",
    "user": {
        "_id": "some_alphanumeric_gibberish",
        "name": "Dwayne Johnson",
        "avatar": ""
    },
    "text": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
    "comments": [
        {
            "user": {
                "_id": "some_alphanumeric_gibberish2",
                "name": "Vin Diesel",
                "avatar": ""
            },
            "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "date": "1535133201"
        },
        {
            "user": {
                "_id": "some_alphanumeric_gibberish2",
                "name": "Vin Diesel",
                "avatar": ""
            },
            "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "date": "1535133201"
        },
        {
            "user": {
                "_id": "some_alphanumeric_gibberish2",
                "name": "Vin Diesel",
                "avatar": ""
            },
            "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "date": "1535133201"
        }
    ],
    "likes": [
        {
            "_id": "19825918265102965",
            "user": "aoeu9a8o10s392h0"
        },
        {
            "_id": "19825918265102966",
            "user": "aoeu9a8o10s392hx0"
        },
        {
            "_id": "19825918265102967",
            "user": "aoeu9a8o10s392120"
        }
    ],
    "date": "1535133024"
}

export default new Array(10).fill(mockPost, 0, 10)
