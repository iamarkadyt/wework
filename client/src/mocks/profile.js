export default {
    "skills": [
        "Assembler",
        "Go",
        "C#",
        "C++"
    ],
    "social": [],
    "date": "2018-08-23T20:25:12.263Z",
    "_id": "5b7b42ba57a5806b896e8f0d",
    "user": {
        "_id": "5b75f023f1957052e52106f5",
        "name": "John Doe"
    },
    "__v": 0,
    "handle": "techguy_handle",
    "status": "Actively applying",
    "title": "Senior Developer",
    "company": "Itron",
    "experience": [
        {
            "current": false,
            "_id": "5b7ca6cf63353349cf03756c",
            "title": "SWE",
            "company": "Revature",
            "from": "2009-03-31T23:00:00.000Z",
            "to": "2025-08-01T01:13:50.111Z"
        }
    ],
    "education": [
        {
            "current": true,
            "_id": "5b7cab633345cc4c066c0540",
            "school": "AC Bootcamp",
            "degree": "Nano CS Degree",
            "from": "2015-08-02T18:33:20.000Z"
        }
    ],
    "bio": "My name is [techguyinfo]. I am a senior developer currently working at Microsoft. I specialize in AI and Cloud computing.",
    "githubusername": "arkadyt",
    "location": "Seattle, WA",
    "website": "techguy.info"
}