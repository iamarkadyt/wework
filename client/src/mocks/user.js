export default {
  id: "98ac61abed72ee8d12"
}
