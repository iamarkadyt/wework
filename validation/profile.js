const Yup = require('yup')
const Profile = require('../models/Profile')

/**
 * Check to see if value (by its pathname) exists in the database.
 * 'user' ObjectID is required to see if it belongs to the current user,
 * in which case there is no error --> user is updating his profile. 
 */
Yup.addMethod(Yup.mixed, 'unoccupied', function (user, message) {
    return this.test('unoccupied', message, function (value) {
        const { path, createError } = this;
        return new Promise((resolve, reject) => {
            // check to see if any profile has already taken the value
            Profile.findOne({ [path]: value }).then(profile => {
                // compare mongoose's ObjectIDs with mongoose's .equals()
                if (profile && !profile.user.equals(user))
                    reject(createError({ path, message }))
                // else value is free, or it belongs to the current user
                else resolve(true)
            })
        })
    })
})

module.exports = (profile, options = {
    onlyEducation: false,
    onlyExperience: false
}) => new Promise((resolve, reject) => {
    let schema = {}

    if (options.onlyExperience) {
        schema = Yup.object().shape({
            title: Yup.string().required("Please specify your title"),
            company: Yup.string().required("Please enter the company name"),
            location: Yup.string(),
            from: Yup.string().required("Please specify the start date"),
            to: Yup.string().when('current', {
                is: undefined,
                then: Yup.string().required("Do you still work here?"),
                otherwise: Yup.string().notRequired()
            }),
            current: Yup.bool(),
            description: Yup.string().min(300, "Please enter at least 300 characters")
        })
    } else if (options.onlyEducation) {
        schema = Yup.object().shape({
            school: Yup.string().required("Please enter a school name"),
            degree: Yup.string().required("Please specify the degree type"),
            fieldOfStudy: Yup.string().required("Please specify your field of study"),
            from: Yup.string().required("Please enter the start date"),
            to: Yup.string().when('current', {
                is: undefined,
                then: Yup.string().required("Do you still study here?"),
                otherwise: Yup.string().notRequired()
            }),
            current: Yup.bool(),
            description: Yup.string().min(300, "Minimum of 300 characters is required!")
        })
    } else { /* For whole profile with the exclusion of above described fields */
        schema = Yup.object().shape({
            handle: Yup.string()
                .required('Please enter your profile handle')
                .min(2, 'A minimum of 2 characters is required')
                .max(40, 'Maximum allowed characters is 40')
                .unoccupied(profile.user, 'Handle is already taken'),
            status: Yup.string()
                .required('Please specify your job seeker status'),
            title: Yup.string()
                .required('Please specify your job title'),
            skills: Yup.array().of(Yup.string().typeError('Elements must be strings'))
                .min(3, 'Please enter at least ${min} skills')
                .max(40, 'Skills limit of ${max} has been reached')
                .required('Please specify your skills'),
            youtube: Yup.string().url('Please enter a valid URL address'),
            twitter: Yup.string().url('Please enter a valid URL address'),
            facebook: Yup.string().url('Please enter a valid URL address'),
            linkedin: Yup.string().url('Please enter a valid URL address'),
            instagram: Yup.string().url('Please enter a valid URL address'),
            githubusername: Yup.string()
                .matches(/^[a-z\d](?:[a-z\d]|-(?=[a-z\d])){0,38}$/i,
                    'Please enter a valid github username')
        })
    }

    // strict flag ensures that value coercion and object transformation would not happen
    // this way db stays clean of empty fields 
    schema.validate(profile, { abortEarly: false, strict: true })
        .then(profile => resolve(profile))
        .catch(err => {
            const errors = {}
            for (let i of err.inner)
                errors[i.path] = i.message
            reject(errors)
        })
})