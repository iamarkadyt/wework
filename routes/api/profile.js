const express = require('express')
const router = express.Router()
const mongoose = require('mongoose')
const passport = require('passport')
const Profile = require('../../models/Profile')
const User = require('../../models/User')

// @route   GET api/profile/test
// @desc    Tests profile route.
// @access  Public
router.get('/test', (req, res) => {
    res.json({ message: 'OK' })
})


// @route   GET api/profile
// @desc    Get current users profile
// @access  Private
router.get('/', passport.authenticate('jwt', { session: false }), (req, res) => {
    const errors = {}

    Profile.findOne({ user: req.user.id })
        .then(profile => {
            if (!profile) {
                errors.noprofile = 'There is no profile for this user'
                res.status(404).json({ errors })
            }

            res.json(profile)
        })
        .catch(err => res.status(404).json(err))
})

// @route   POST /api/profile
// @desc    Create or Update user profile
// @access  Protected
router.post('/', passport.authenticate('jwt', { session: false }), (req, res) => {
    require('../../validation/profile')({ ...req.body, user: req.user.id })
        .then(profile => {
            Profile.findOneAndUpdate(
                { user: req.body.id },
                { $set: profile }, // form fields left empty must not be sent
                { new: true, upsert: true })
                .then(profile => res.json(profile))
                .catch(err => res.status(400).json(err))
        })
        .catch(errors => res.status(400).json(errors))
})


module.exports = router;