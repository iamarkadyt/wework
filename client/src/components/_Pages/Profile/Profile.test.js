import React from 'react'
import { Profile } from './Profile'
import { shallow } from 'enzyme'
import { BrowserRouter } from 'react-router-dom'
import cloneDeep from 'lodash.clonedeep'
import { mockProfile } from '../../../mocks/profile'

let mountedComponent, props

const getMockProps = () => {
  return cloneDeep({
    history: {
      goBack: jest.fn(),
      push: jest.fn()
    },
    match: {
      url: '/profile'
    },
    fetchUsersProfile: jest.fn(),
    deleteAProfile: jest.fn(),
    fetchAProfile: jest.fn(),
    viewedProfile: mockProfile,
    profile: mockProfile,
    errors: {}
  })
}

const createRouterContext = pathname => {
  return cloneDeep({
    router: {
      history: new BrowserRouter().history,
      route: {
        location: {
          pathname
        },
        match: {},
      }
    }
  })
}

const comp = () => {
  if (!mountedComponent) {
    mountedComponent = shallow(<Profile {...props} />)
  } 
  return mountedComponent
}

describe('Profile', () => {
  beforeEach(() => {
    mountedComponent = undefined
    props = getMockProps()
  })

  describe('snapshots', () => {
    describe('Routes', () => {
      describe('rendered when authed user is at /profile, and he/she already have a profile', () => {
        beforeEach(() => {
          props.profile = mockProfile
          delete props.errors.noProfile
        })

        it('/delete-profile', () => {
          const context = createRouterContext(`${props.match.url}/delete-profile`)
          expect(comp().find('Route[path*="delete-profile"]').dive({ context })).toMatchSnapshot()
        })

        it('/add-education', () => {
          const context = createRouterContext(`${props.match.url}/add-education`)
          expect(comp().find('Route[path*="add-education"]').dive({ context })).toMatchSnapshot()
        })

        it('/add-experience', () => {
          const context = createRouterContext(`${props.match.url}/add-experience`)
          expect(comp().find('Route[path*="add-experience"]').dive({ context })).toMatchSnapshot()
        })

        it('/update-profile', () => {
          const context = createRouterContext(`${props.match.url}/update-profile`)
          expect(comp().find('Route[path*="update-profile"]').dive({ context })).toMatchSnapshot()
        })

        it('bare baseUrl', () => {
          const context = createRouterContext(`${props.match.url}`)
          expect(comp().find(`Route[path="${props.match.url}"]`).dive({ context })).toMatchSnapshot()
        })
      })

      describe('rendered when authed user is at /profile, and has no profile yet', () => {
        beforeEach(() => {
          delete props.profile
          props.errors.noProfile = 'defined'
        })

        it('/create-profile', () => {
          const context = createRouterContext(`${props.match.url}/create-profile`)
          expect(comp().find('Route[path*="create-profile"]').dive({ context })).toMatchSnapshot()
        })

        it('bare baseUrl', () => {
          const context = createRouterContext(`${props.match.url}`)
          expect(comp().find(`Route[path="${props.match.url}"]`).dive({ context })).toMatchSnapshot()
        })
      })

      describe('rendered when authed user is at /profile/id, and viewed user has a profile', () => {
        beforeEach(() => {
          props.viewedProfile = mockProfile
        })
      })
    })

    describe('conditional rendering outputs', () => {

    })
  })
})

