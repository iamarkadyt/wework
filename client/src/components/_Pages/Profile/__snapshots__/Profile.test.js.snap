// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile snapshots Routes rendered when authed user is at /profile, and has no profile yet /create-profile 1`] = `
<Connect(UpdateProfile)
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "pathname": "/profile/create-profile",
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/profile/create-profile",
      "url": "/profile/create-profile",
    }
  }
/>
`;

exports[`Profile snapshots Routes rendered when authed user is at /profile, and has no profile yet bare baseUrl 1`] = `
<Modal
  actionColor="green"
  onConfirm={[Function]}
  onDismiss={[Function]}
  question="You don't yet have a profile. Would you like to create one now?"
/>
`;

exports[`Profile snapshots Routes rendered when authed user is at /profile, and he/she already have a profile /add-education 1`] = `
<Connect(AddEdu)
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "pathname": "/profile/add-education",
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/profile/add-education",
      "url": "/profile/add-education",
    }
  }
/>
`;

exports[`Profile snapshots Routes rendered when authed user is at /profile, and he/she already have a profile /add-experience 1`] = `
<Connect(AddExp)
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "pathname": "/profile/add-experience",
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/profile/add-experience",
      "url": "/profile/add-experience",
    }
  }
/>
`;

exports[`Profile snapshots Routes rendered when authed user is at /profile, and he/she already have a profile /delete-profile 1`] = `
<Modal
  actionColor="red"
  onConfirm={[Function]}
  onDismiss={[Function]}
  question="Are you sure you want to delete your profile?"
/>
`;

exports[`Profile snapshots Routes rendered when authed user is at /profile, and he/she already have a profile /update-profile 1`] = `
<Connect(UpdateProfile)
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "pathname": "/profile/update-profile",
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/profile/update-profile",
      "url": "/profile/update-profile",
    }
  }
  profile={
    Object {
      "__v": 0,
      "_id": "5b7b42ba57a5806b896e8f0d",
      "bio": "TEST BIO TEXT.",
      "company": "Some Company",
      "date": "2018-08-23T20:25:12.263Z",
      "education": Array [
        Object {
          "_id": "5b7cab633345cc4c066c0540",
          "current": true,
          "degree": "Nano CS Degree",
          "fieldOfStudy": "Bioengineering",
          "from": "2015-08-02T00:00:00.000Z",
          "school": "Some Bootcamp",
        },
      ],
      "experience": Array [
        Object {
          "_id": "5b7ca6cf63353349cf03756c",
          "company": "Some Company",
          "current": false,
          "from": "2009-03-31T23:00:00.000Z",
          "title": "Software Developer",
          "to": "2012-08-01T00:00:00.000Z",
        },
      ],
      "githubusername": "arkadyt",
      "handle": "john_doe",
      "location": "Seattle, WA",
      "skills": Array [
        "Go",
        "C#",
        "C++",
        "Whatnot else",
      ],
      "status": "Not available for employment",
      "title": "Senior Developer",
      "user": Object {
        "_id": "5b75f023f1957052e52106f5",
        "avatar": "gravatar://some.avatar",
        "name": "John Doe",
      },
      "website": "arkadyt.com",
    }
  }
/>
`;

exports[`Profile snapshots Routes rendered when authed user is at /profile, and he/she already have a profile bare baseUrl 1`] = `
<Connect(ProfileView)
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "pathname": "/profile",
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/profile",
      "url": "/profile",
    }
  }
  profile={
    Object {
      "__v": 0,
      "_id": "5b7b42ba57a5806b896e8f0d",
      "bio": "TEST BIO TEXT.",
      "company": "Some Company",
      "date": "2018-08-23T20:25:12.263Z",
      "education": Array [
        Object {
          "_id": "5b7cab633345cc4c066c0540",
          "current": true,
          "degree": "Nano CS Degree",
          "fieldOfStudy": "Bioengineering",
          "from": "2015-08-02T00:00:00.000Z",
          "school": "Some Bootcamp",
        },
      ],
      "experience": Array [
        Object {
          "_id": "5b7ca6cf63353349cf03756c",
          "company": "Some Company",
          "current": false,
          "from": "2009-03-31T23:00:00.000Z",
          "title": "Software Developer",
          "to": "2012-08-01T00:00:00.000Z",
        },
      ],
      "githubusername": "arkadyt",
      "handle": "john_doe",
      "location": "Seattle, WA",
      "skills": Array [
        "Go",
        "C#",
        "C++",
        "Whatnot else",
      ],
      "status": "Not available for employment",
      "title": "Senior Developer",
      "user": Object {
        "_id": "5b75f023f1957052e52106f5",
        "avatar": "gravatar://some.avatar",
        "name": "John Doe",
      },
      "website": "arkadyt.com",
    }
  }
/>
`;
